File: src/Core/TagReview.cs
using System.Security.Cryptography;
namespace Omni.Core;

public static class TagReview
{
    public static readonly (string Id,string Label)[] Fields = [
        ("TIT2","標題"),("TPE1","演出者"),("TALB","專輯"),("TPE2","專輯演出者"),
        ("TRCK","曲目"),("TPOS","光碟"),("TYER","年份"),("TDOR","原始發行日期"),
        ("TCON","類型"),("TCOM","作曲者"),("TPUB","發行商標"),("TSRC","ISRC"),
        ("TXXX:MusicBrainz Recording Id","MusicBrainz 錄音 ID"),("TXXX:MusicBrainz Album Id","MusicBrainz 發行 ID"),
        ("TSOT","標題排序方式"),("TSOP","演出者排序方式")];
    public static string Frame(string id,byte version)=>id=="TYER"&&version==4?"TDRC":id=="TDOR"&&version==3?"TORY":id;
    public static Dictionary<string,string> Values(Id3Document doc)=>Fields.ToDictionary(f=>f.Id,f=>doc.Text(Frame(f.Id,doc.Version)));
    public static async Task<string> Hash(string path){await using var stream=File.OpenRead(path);return Convert.ToHexString(await SHA256.HashDataAsync(stream));}
}

public sealed class TagReviewUndo
{
    public string Backup {get;}
    readonly string path,afterHash; readonly DownloadJob before;
    TagReviewUndo(string backup,string path,string hash,DownloadJob before){Backup=backup;this.path=path;afterHash=hash;this.before=before;}
    public static async Task<TagReviewUndo?> Apply(Store store,DownloadJob job,Dictionary<string,string> selected,Cover? cover,bool keep,string originalHash){
        var path=job.FilePath??throw new IOException("找不到檔案");
        if(await TagReview.Hash(path)!=originalHash)throw new IOException("檔案已被其他操作修改，請返回編輯再掃描。");
        var before=Json.Decode<DownloadJob>(Json.Encode(job));var directory=Path.Combine(store.DataDirectory,"tag-undo");Directory.CreateDirectory(directory);
        var backup=Path.Combine(directory,Guid.NewGuid()+".mp3");if(keep)File.Copy(path,backup);
        try{var doc=Id3Document.Read(path);var mapped=selected.ToDictionary(p=>TagReview.Frame(p.Key,doc.Version),p=>p.Key=="TDOR"&&doc.Version==3&&p.Value.Length>=4?p.Value[..4]:p.Value);
            await new TagEditor(store).Apply([job],new(mapped,Covers:cover is null?null:[cover],RenameFile:false),store.Load().Any(j=>j.Id==job.Id),expectedHash:originalHash);
            if(!keep)return null;var undo=new TagReviewUndo(backup,path,await TagReview.Hash(path),before);
            File.WriteAllText(backup+".json",Json.Encode(new{Path=path,Before=before,AfterHash=undo.afterHash}));return undo;
        }catch{if(File.Exists(backup))File.Delete(backup);throw;}
    }
    public async Task Restore(Store store,DownloadJob job){
        if(await TagReview.Hash(path)!=afterHash)throw new IOException("檔案其後已有修改，為避免覆蓋，不能直接復原。");
        var stage=path+".undo-"+Guid.NewGuid();File.Copy(Backup,stage);try{File.Move(stage,path,true);}finally{if(File.Exists(stage))File.Delete(stage);}
        job.Title=before.Title;job.Artist=before.Artist;job.Album=before.Album;job.IsUserEdited=before.IsUserEdited;job.CoverUserEdited=before.CoverUserEdited;
        if(store.Load().Any(j=>j.Id==job.Id))store.Save(job);File.Delete(Backup);File.Delete(Backup+".json");
    }
}

File: android/app/src/main/java/io/hkmario/omni/TagReview.kt
package io.hkmario.omni
import android.net.Uri
import kotlinx.coroutines.*
import java.io.File
import java.util.UUID
import java.security.MessageDigest

object AcoustIdClient { fun resolve(custom:String)=custom.trim().ifBlank { "wdBJF1kUQS" } }
object TagReview {
 val fields=listOf("TIT2" to "標題","TPE1" to "演出者","TALB" to "專輯","TPE2" to "專輯演出者","TRCK" to "曲目","TPOS" to "光碟","TYER" to "年份","TDOR" to "原始發行日期","TCON" to "類型","TCOM" to "作曲者","TPUB" to "發行商標","TSRC" to "ISRC","TXXX:MusicBrainz Recording Id" to "MusicBrainz 錄音 ID","TXXX:MusicBrainz Album Id" to "MusicBrainz 發行 ID","TSOT" to "標題排序方式","TSOP" to "演出者排序方式")
 fun input(engine:Engine,path:String)=if(path.startsWith("content://"))engine.context.contentResolver.openInputStream(Uri.parse(path))?:error("檔案權限已失效")else File(path).inputStream()
 fun hash(engine:Engine,path:String):String {val digest=MessageDigest.getInstance("SHA-256");input(engine,path).use{val b=ByteArray(65536);while(true){val n=it.read(b);if(n<0)break;digest.update(b,0,n)}};return digest.digest().joinToString(""){"%02x".format(it)}}
 suspend fun read(engine:Engine,song:TaskItem)=withContext(Dispatchers.IO){val source=File(engine.context.cacheDir,"review-${UUID.randomUUID()}.mp3");try{input(engine,song.path!!).use{i->source.outputStream().use{i.copyTo(it)}};val d=Id3.read(source);SongTags(fields.associate{(id,_)->id to d.text(if(id=="TYER"&&d.version==4)"TDRC"else if(id=="TDOR"&&d.version==3)"TORY"else id)},d.artwork())}finally{source.delete()}}
 suspend fun apply(engine:Engine,song:TaskItem,values:Map<String,String>,cover:ByteArray?,keep:Boolean,originalHash:String):ReviewUndo?=withContext(Dispatchers.IO+NonCancellable){
  require(hash(engine,song.path!!)==originalHash){"檔案已被修改，請返回編輯再掃描"}
  val backup=File(engine.context.filesDir,"tag-undo/${UUID.randomUUID()}.mp3");if(keep){backup.parentFile!!.mkdirs();input(engine,song.path).use{i->backup.outputStream().use{i.copyTo(it)}}}
  try{TagEditor(engine).apply(setOf(song.id),values,emptyMap(),cover,renameFile=false,expectedHash=originalHash);if(!keep)null else ReviewUndo(backup,song,hash(engine,song.path)).also{File(backup.path+".json").writeText(org.json.JSONObject().put("path",song.path).put("afterHash",it.after).toString())}}catch(e:Exception){backup.delete();throw e}
 }
}
data class ReviewUndo(val backup:File,val before:TaskItem,val after:String){
 suspend fun restore(engine:Engine)=withContext(Dispatchers.IO+NonCancellable){val path=before.path!!;require(TagReview.hash(engine,path)==after){"檔案其後已有修改，不能直接復原"};if(path.startsWith("content://")){val recovery=File(engine.context.cacheDir,"undo-recovery-${UUID.randomUUID()}.mp3");TagReview.input(engine,path).use{i->recovery.outputStream().use{i.copyTo(it)}};try{engine.context.contentResolver.openOutputStream(Uri.parse(path),"wt")!!.use{o->backup.inputStream().use{it.copyTo(o)}}}catch(e:Exception){runCatching{engine.context.contentResolver.openOutputStream(Uri.parse(path),"wt")!!.use{o->recovery.inputStream().use{it.copyTo(o)}}};throw e}finally{recovery.delete()}}else{val stage=File(path+".undo-${UUID.randomUUID()}");backup.copyTo(stage);java.nio.file.Files.move(stage.toPath(),File(path).toPath(),java.nio.file.StandardCopyOption.REPLACE_EXISTING)};engine.update(before.id){before};backup.delete();File(backup.path+".json").delete()}
}
